Library/Homebrew/livecheck/strategy/github_latest.rb

```diff
@@ -0,0 +1,73 @@
+# typed: false
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      # The {GithubLatest} strategy identifies versions of software at
+      # github.com by checking a repository's "latest" release page.
+      #
+      # GitHub URLs take a few different formats:
+      #
+      # * `https://github.com/example/example/releases/download/1.2.3/example-1.2.3.tar.gz`
+      # * `https://github.com/example/example/archive/v1.2.3.tar.gz`
+      # * `https://github.com/downloads/example/example/example-1.2.3.tar.gz`
+      #
+      # A repository's `/releases/latest` URL normally redirects to a release
+      # tag (e.g., `/releases/tag/1.2.3`). When there isn't a "latest" release,
+      # it will redirect to the `/releases` page.
+      #
+      # This strategy should only be used when we know the upstream repository
+      # has a "latest" release and the tagged release is appropriate to use
+      # (e.g., "latest" isn't wrongly pointing to an unstable version, not
+      # picking up the actual latest version, etc.). The strategy can only be
+      # applied by using `strategy :github_latest` in a `livecheck` block.
+      #
+      # The default regex identifies versions like `1.2.3`/`v1.2.3` in `href`
+      # attributes containing the tag URL (e.g.,
+      # `/example/example/releases/tag/v1.2.3`). This is a common tag format
+      # but a modified regex can be provided in a `livecheck` block to override
+      # the default if a repository uses a different format (e.g.,
+      # `example-1.2.3`, `1.2.3d`, `1.2.3-4`, etc.).
+      #
+      # @api public
+      class GithubLatest
+        NICE_NAME = "GitHub - Latest"
+
+        # A priority of zero causes livecheck to skip the strategy. We do this
+        # for {GithubLatest} so we can selectively apply the strategy using
+        # `strategy :github_latest` in a `livecheck` block.
+        PRIORITY = 0
+
+        # The `Regexp` used to determine if the strategy applies to the URL.
+        URL_MATCH_REGEX = %r{//github\.com(?:/downloads)?(?:/[^/]+){2}}i.freeze
+
+        # Whether the strategy can be applied to the provided URL.
+        #
+        # @param url [String] the URL to match against
+        # @return [Boolean]
+        def self.match?(url)
+          URL_MATCH_REGEX.match?(url)
+        end
+
+        # Generates a URL and regex (if one isn't provided) and passes them
+        # to {PageMatch.find_versions} to identify versions in the content.
+        #
+        # @param url [String] the URL of the content to check
+        # @param regex [Regexp] a regex used for matching versions in content
+        # @return [Hash]
+        def self.find_versions(url, regex = nil)
+          %r{github\.com/(?:downloads/)?(?<username>[^/]+)/(?<repository>[^/]+)}i =~ url.sub(/\.git$/i, "")
```

Comment:
I checked Homebrew/core, and the following packages use `/downloads/` subfolder:

- apktool.rb
- btpd.rb
- cssembed.rb
- growly.rb
- henplus.rb
- jadx.rb
- midgard2.rb
- rpg.rb
- syck.rb
- tidyp.rb
- wrk-trello.rb

My understanding is that `/downloads` folders have been deprecated for a while now -- see https://github.blog/2012-12-12-goodbye-uploads/. I wonder if we should remove `downloads` from here and update the above formulae to use proper download paths.

Reply:
I've already been through the `/downloads/` formulae recently (as part of another PR) and I took care of the straightforward ones, so this is what remains. As you would guess, most of these are highly outdated and tend to have low install counts in our analytics. The repository uploads URLs continue to resolve, so updating these is low priority in comparison to work that can have a greater impact.

## Can potentially be updated

* `apktool.rb`: The `/downloads/` URL is a resource from another GitHub repository that's used in the `test` block but I didn't see an alternative location for this particular file. The formula could be updated to use a different `apk` if a suitable replacement can be found.
* `btpd.rb`: The formula would need to be updated to build using the `0.16` tag archive.
* `jadx.rb`: This uses the same `apk` file as the `apktool` formula, so the situation is the same.
* `midgard2.rb`: The formula would need to be updated to build using the `12.09.1` tag archive.
* `rpg.rb`: The formula would need to be updated to build using the `0.3.0` tag archive. This release is from 2010, the repository is archived (I deprecated the formula in the past), and there have only been 12 installs over the past year, so it may not be worth the effort.
* `tidyp.rb`: The formula would need to be updated to build using the `1.04` tag archive.

## Unlikely that we can update

* `cssembed.rb`: `stable` is a `jar` file and I don't see alternatives. The GitHub repository is archived (I deprecated the formula in the past), so it's not worth the effort.
* `growly.rb`: The repository doesn't have any tags or releases (and we're not going to use the `growly` file from `master`). Besides that, the code hasn't been modified in nine years and the repository doesn't have a license file (ryankee/growly#7).
* `henplus.rb`: The repository doesn't have any tags or releases and the formula is deprecated since it doesn't build. There's been *some* activity here and there in recent years but we're stuck unless they create a release.
* `syck.rb`: This is one of the two formulae using a `github.s3.amazonaws.com/downloads` URL. There aren't any alternatives (to my knowledge) and it hasn't been updated in years (it's a project by why the lucky stiff and I don't imagine him coming back anytime soon).
* `wrk-trello.rb`: This is one of the two formulae using a `github.s3.amazonaws.com/downloads` URL. There aren't any alternatives (to my knowledge) and it hasn't been updated in eight years.